### PR TITLE
feat: Update pre-commit mirror script

### PR DESCRIPTION
**Pull request by sudosubin**
## Summary

- `pre-commit-mirror-maker` is not suitable for generating multiple hooks in single `.pre-commit-hooks.yaml` file.
- So replaced the script and github workflows.

## Test Plan

- Update `pyproject.toml`'s `ruff==0.0.290` into `ruff==0.0.289`, and run the script.

```sh
$ python3 mirror.py
...
```

### CODE REVIEW

**Comment by charliermarsh** [comment]
Nice, thank you so much! We'd just discussed doing this internally, you read our minds... I will defer to @zanieb for review. (Zanie's out on vacation so will probably get to this next week.)

File: mirror.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        raise RuntimeError("Failed to fetch versions from pypi")
+
+    versions = [Version(release) for release in response.json()["releases"]]
+    return sorted(versions)
+
+
+def get_current_version(pyproject: dict) -> Version:
+    requirements = [Requirement(d) for d in pyproject["project"]["dependencies"]]
+    requirement = next((r for r in requirements if r.name == "ruff"), None)
+    assert requirement is not None
```
Comment (zanieb) [inline mirror.py:39]: Should we include an assertion message that would help us diagnose this failed assertion?

File: update_version.py [hunk 2]
```diff
@@ -1,26 +0,0 @@
...
-    readme_md = Path("README.md")
-    readme = readme_md.read_text()
-    rev = Path(".version").read_text().strip()
-    readme = re.sub(r"rev: v\d+\.\d+\.\d+", f"rev: v{rev}", readme)
-    readme = re.sub(r"/ruff/\d+\.\d+\.\d+\.svg", f"/ruff/{rev}.svg", readme)
-    readme_md.write_text(readme)
-
-    # Only commit on change.
-    # https://stackoverflow.com/a/9393642/3549270
-    if check_output(["git", "status", "-s"]).strip():
```
Comment (zanieb) [inline update_version.py:18]: Is this behavior retained?

Comment (sudosubin) [reply update_version.py:18]: Yes! I updated to retain this behavior.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -14,24 +14,21 @@ jobs:
...
 
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
 
       - uses: actions/setup-python@v4
         with:
           python-version: "3.x"
 
-      - run: pip install pre-commit-mirror-maker
+      - run: pip install "urllib3>=2" tomli packaging
```
[the remark this answers is not in the document]
Comment (sudosubin) [reply .github/workflows/main.yml]: I listed dependencies through `optional-dependencies` in `pyproject.toml` and create a `requirements-dev.txt` file through `pip-tools`.

The `setup-python` action now installs the packages via the generated `requirements-dev.txt` file and runs the `mirror.py` script.

File: mirror.py (hunk 1 above)
Comment (sudosubin) [reply mirror.py]: I added some assertion messages, but not sure if I need to supplement.

**Review by zanieb — APPROVED** [review]
Thanks! This looks great

**Pull request merged**